google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2472,38 +2466,79 @@ private void releaseSession(PooledSession session, boolean isNewSession) {
+   *       <ol>
+   *         <li>The first numChannels sessions in the list of sessions contains more than 2
+   *             sessions that use the same channel as the one being added.
+   *         <li>The list of currently checked out sessions contains more than 2 times the the
+   *             number of sessions with the same channel as the one being added than it should in
+   *             order for it to be perfectly balanced. Perfectly balanced in this case means that
+   *             the list should preferably contain size/numChannels sessions of each channel.
+   *       </ol>
+   * </ol>
+   *
+   * @param channelOfSessionBeingAdded the channel number being used by the session that is about to
+   *     be released into the pool
+   * @param sessions the list of all sessions in the pool
+   * @param checkedOutSessions the currently checked out sessions of the pool
+   * @param numChannels the number of channels in use
+   * @return true if the pool is considered unbalanced, and false otherwise
+   */
+  @VisibleForTesting
+  static boolean isUnbalanced(
+      int channelOfSessionBeingAdded,
+      List<PooledSession> sessions,
+      Set<PooledSessionFuture> checkedOutSessions,
+      int numChannels) {
+    // Do not re-balance the pool if the number of checked out sessions is low, as it is
     // better to re-use sessions as much as possible in a low-QPS scenario.
     if (sessions.isEmpty() || checkedOutSessions.size() <= 2) {
       return false;
     }
-    int numChannels = sessionClient.getSpanner().getOptions().getNumChannels();
     if (numChannels == 1) {
       return false;
     }
 
     // Ideally, the first numChannels sessions in the pool should contain exactly one session for
     // each channel.
     // Check if the first numChannels sessions at the head of the pool already contain more than 2
-    // sessions that use the same channel as this one. If so, we randomize.
-    int channel = session.getChannel();
+    // sessions that use the same channel as this one. If so, we re-balance.
+    // We also re-balance the pool in the specific case that the pool uses 2 channels and the first
+    // two sessions use those two channels.
+    int maxSessionsAtHeadOfPool = Math.min(numChannels, 3);
     int count = 0;
     for (int i = 0; i < Math.min(numChannels, sessions.size()); i++) {
       PooledSession otherSession = sessions.get(i);
-      if (channel == otherSession.getChannel()) {
+      if (channelOfSessionBeingAdded == otherSession.getChannel()) {
         count++;
-        if (count > 2) {
+        if (count >= maxSessionsAtHeadOfPool) {
           return true;
         }
       }
     }
     // Ideally, the use of a channel in the checked out sessions is exactly
     // numCheckedOut / numChannels
     // We check whether we are more than a factor two away from that perfect distribution.
-    // If we are, then we randomize.
+    // If we are, then we re-balance.
+    count = 0;
```

Comment:
This was a bug in the previous version that was found with the additional test cases.